akka-stream/src/main/scala/akka/stream/impl/fusing/FutureFlow.scala

```diff
@@ -0,0 +1,170 @@
+/*
+ * Copyright (C) 2020 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.stream.impl.fusing
+
+import akka.annotation.InternalApi
+import akka.dispatch.ExecutionContexts
+import akka.stream.{ AbruptStageTerminationException, Attributes, FlowShape, Inlet, NeverMaterializedException, Outlet }
+import akka.stream.scaladsl.{ Flow, Keep, Source }
+import akka.stream.stage.{ GraphStageLogic, GraphStageWithMaterializedValue, InHandler, OutHandler }
+import akka.util.OptionVal
+
+import scala.concurrent.{ Future, Promise }
+import scala.util.{ Failure, Success, Try }
+
+@InternalApi private[akka] final class FutureFlow[In, Out, M](future: Future[Flow[In, Out, M]])
+    extends GraphStageWithMaterializedValue[FlowShape[In, Out], Future[M]] {
+  val in = Inlet[In](s"${this}.in")
+  val out = Outlet[Out](s"${this}.out")
+  override val shape: FlowShape[In, Out] = FlowShape(in, out)
+
+  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes): (GraphStageLogic, Future[M]) = {
+    val innerMatValue = Promise[M]
+    val logic = new GraphStageLogic(shape) {
+
+      //seems like we must set handlers BEFORE preStart
+      setHandlers(in, out, initializing)
+
+      override def preStart(): Unit = {
+        super.preStart()
+        future.value match {
+          case Some(tr) =>
+            initializing.onFuture(tr)
+          case None =>
+            val cb = getAsyncCallback(initializing.onFuture)
+            future.onComplete(cb.invoke)(ExecutionContexts.parasitic)
+            //in case both ports are closed before future completion
+            setKeepGoing(true)
+        }
+      }
+
+      override def postStop(): Unit = {
+        if (!innerMatValue.isCompleted) {
+          innerMatValue.failure(new AbruptStageTerminationException(this))
+        }
+        super.postStop()
+      }
+
+      object initializing extends InHandler with OutHandler {
+        override def onPush(): Unit = {
+          throw new IllegalStateException("unexpected push during initialization")
+        }
+
+        var upstreamFailure = OptionVal.none[Throwable]
+        override def onUpstreamFailure(ex: Throwable): Unit = {
+          upstreamFailure = OptionVal.Some(ex)
+        }
+
+        override def onUpstreamFinish(): Unit = {
+          //will later be propagated to the materialized flow
```

Comment:
Shouldn't this be handled during initialization too?

Reply:
it is, it just doesn't require any specific state on top of `isClosed(in)`

Reply:
I'm not sure I understand.  How do we close the stage when its in `initializing` when the upstream completes?

Reply:
when initializing stage set's keepGoing=true, then when the future completes the stage changes state, materializes the flow, connects its ports and forwards the signals.
specifically it checks isClosed(in) and closes the subSource in such cases.

Reply:
I'm not the signals would be forwarded after new handlers are set.  `onUpstreamFinish` will only be called once, so if that happens before the future completes then we'll miss the signal.

Reply:
when onUpstreamFinish is called before future completion we do not close the stage, furthermore even if downstream is completed we do **not** close the stage (due to `setKeepGoing(true)`).
at some point the future completes and the flow is materialized. at this point the 'stashed' signals are forwarded to the materialized flow, this may include an upstream completion or failure, and downstream pull cancellation or failure.

another related point, it is possible for the 'outer' stage to complete while the 'inner' flow keeps running. this is achieved by `setKeepGoinf(false)` after materializing the nested flow and forwarding the signals, if both upstream and downstream are closed at this point the outer stage will complete but it's up to the nested flow to decide how to handle the closed ports. you can see examples in the tests.

Reply:
> at some point the future completes and the flow is materialized. at this point the 'stashed' signals are forwarded to the materialized flow, this may include an upstream completion or failure, and downstream pull cancellation or failure.

I'm unfamiliar with this behaviour, but I'll dig into it for my own understanding.  I'll follow up.

Reply:
@seglo, see line 120 - 126. The stashing is part manual / part built-in. If there's a failure it's explicitly saved in `upstreamFailure`. If it has been completed, `isClosed(in)` will return true. So, no magic involved.

Reply:
thanks @jrudolph, @seglo notice this code is now migrated into the connect method you guys requested.

Reply:
Thanks for the clarification.  I see it now.  Thanks.